packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
+        builder.addBlob(".deletedDataStores", JSON.stringify(deletedDataStores));
+
         this.serializeContainerBlobs(builder);
         const summary = builder.getSummaryTree();
         return { ...summary, id: "" };
     }
 
+    public runGarbageCollection(): string[] {
+        // Maintains a map of dataStoreId to an IReferences object. The IReferences object keeps track of whether
+        // the data store has been visited (to detect loops) and the fluid objects in it that are referenced.
+        const dataStoreReferencesMap: Map<string, IReferences> = new Map();
+
+        // A stack of routes that are referenced through handles in DDSes.
+        const referencedRoutesStack: string[] = [];
+
+        // Start with the root data stores. Add all the fluid objects referenced by the root data stores in to the
+        // referencedObjectsStack.
+        this.contexts.forEach((value: FluidDataStoreContext) => {
+            if (value.isRootStore) {
+                // Ignore the scheduler.
+                if (value.id === taskSchedulerId) {
+                    return;
+                }
+
+                // Get the fluid objects referenced by this data store and add them to the referencedObjectsStack.
+                const referencedRoutes = value.getReferencedRoutes();
+                referencedRoutesStack.push(...referencedRoutes);
+
+                // Set the IReferences object in the dataStoreReferencesMap for this data store. Mark it as visited.
+                dataStoreReferencesMap.set(value.id, { visited: true, referencedRoutes: new Set<string>() });
+            }
+        });
+
+        // Do a Bread First Search on the referencedRoutesStack.
+        while (referencedRoutesStack.length > 0) {
+            // Get the next route.
+            const nextRoute = referencedRoutesStack.shift();
+            assert(nextRoute);
+
+            // Get the dataStoreId from the referenced object.
+            const dataStoreId = nextRoute.split("/")[1];
+            // If there is no reference object for it in the dataStoreReferencesMap, it means this is the first time we
+            // are seeing it. Creata a reference object for it.
+            let references = dataStoreReferencesMap.get(dataStoreId);
+            if (references === undefined) {
+                references = {
+                    visited: false,
+                    referencedRoutes: new Set<string>(),
+                };
+            }
+
+            const visited: boolean = references.visited;
+            // Keeps track of whether the entry for this data store in the dataStoreReferencesMap changes.
+            let dataStoreEntryChanged: boolean = false;
+
+            // Get the path of the referenced object relative to the dataStoreId and add it to the
+            // referencedFluidObjects in its references.
+            const dataStoreIdRelativePath = nextRoute.slice(dataStoreId.length + 1);
+            if (dataStoreIdRelativePath !== "") {
```

Comment:
I think (but not sure) that this results in never adding "/" to the list, i.e. path that does not go deeper (default path)

Reply:
We don't need to add it to the list right. The list keeps track of all child objects withing a data store that are referenced. So, we don't need to add "/" to the list.
It is already part of the "referencedRoutesStack" which keeps track of all referenced data stores.

Reply:
/ is also a child object (in current system). I.e. I can have reference of type /ID/ and that refers to default object of data store - data store needs to know about it.

Reply:
Ah okay. I will add code to handle this.